examples/user_profiles/connector.py

```diff
@@ -0,0 +1,150 @@
+# The schema function takes one parameter:
+# - configuration: a dictionary that holds the configuration settings for the connector.
+def schema(configuration: dict):
+    return [
+        {
+            "table": "profile",
+            "primary_key": ["id"],
+            # Columns and data types will be inferred by Fivetran
+        },
+        {
+            "table": "location",
+            "primary_key": ["profileId"],
+            # Columns and data types will be inferred by Fivetran
+        },
+        {
+            "table": "login",
+            "primary_key": ["profileId"],
+            # Columns and data types will be inferred by Fivetran
+        }
+    ]
+
+# Define the update function, which is a required function, and is called by Fivetran during each sync.
+# See the technical reference documentation for more details on the update function:
+# https://fivetran.com/docs/connectors/connector-sdk/technical-reference#update
+# The function takes two parameters:
+# - configuration: dictionary containing any secrets or payloads you configure when deploying the connector.
+# - state: a dictionary containing the state checkpointed during the prior sync.
+#   The state dictionary is empty for the first sync or for any full re-sync.
+def update(configuration: dict, state: dict):
+    # Retrieve the last processed profile ID from state or set it to 0 if not present
+    profile_cursor = state["profile_cursor"] if "profile_cursor" in state else 0
+
+    # Fetch new data and return DataFrames for profile, location, and login tables
+    profile_df, location_df, login_df, cursor = get_data(profile_cursor)
+
+
+    # Approaches to upsert the records from dataFrame
+
+    # APPROACH - 1
+    # UPSERT all profile table data, checkpoint periodically to save progress. In this example every 5 records.
+    for index, row in profile_df.iterrows():
+        yield op.upsert("profile", {col: row[col] for col in profile_df.columns})
+        if index % 5 == 0:
+            state["profile_cursor"] = row["id"]
+            yield op.checkpoint(state)
+    
+    # Checkpointing at the end of the "profile" table data processing
+    state["profile_cursor"] = cursor
+    yield op.checkpoint(state)
+
+    # APPROACH - 2
+    # UPSERT all location table data.
+    # Iterate over each row in the DataFrame, converting it to a dictionary
+    for row in location_df.to_dict("records"):
+        yield op.upsert("location", row)
+
+    # Checkpointing at the end of the "location" table data processing
+    state["location_cursor"] = cursor
+    yield op.checkpoint(state)
+
+    # APPROACH - 3
+    # UPSERT all login table data.
+    # Iterate over the values of the dictionary (which are the DataFrame rows)
+    for value in login_df.to_dict("index").values():
+        yield op.upsert("login", value)
+    
+    # Checkpointing at the end of the "login" table data processing
+    state["login_cursor"] = cursor
+    yield op.checkpoint(state)
+
```

Comment:
Should we give some indication of when to use which approach?
are their any pros/cons for the different approaches - is one faster than the others?

Reply:
@alsnkline Added some information about the approaches in the comments, please let me know if it looks good.